share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+
+> **_Note:_** If `MaximumGPRsPerScalarArgument` is equal to 0, it means that the ABI does not allow scalar arguments to use GPRs.
+
+> **_Note:_** If an argument doesn't fit into the specified registers or uses irregular padding, the registers are not used and the object is passed using the memory (stack, pointer-to-copy, etc.).
+
+## `MaximumGPRsPerScalarReturnValue` (`unsigned`)
+
+States the maximum number of GPRs available to return a single scalar (`int`, `__int128`, pointer, etc.) value, meaning that it can only be returned in the GPRs if `MaximumGPRsPerScalarReturnValue` is less than or equal to the number of the registers required to fit the object including padding.
+
+> **_Note:_** If `MaximumGPRsPerScalarReturnValue` is equal to 0, it means that the ABI does not allow scalar return values to use GPRs.
+
+> **_Note:_** If a return value doesn't fit into the specified registers or uses irregular padding, the registers are not used and the object is passed using the memory (stack, pointer-to-copy, etc.).
+
+## `GeneralPurposeArgumentRegisters` (`list of registers`)
+
+Specifies the list of general purpose registers allowed to be used for passing arguments and the order they are to be used in.
+
+## `GeneralPurposeReturnValueRegisters` (`list of registers`)
+
+Specifies the list of general purpose registers allowed to be used for returning values and the order they are to be used in.
+
+## `VectorArgumentRegisters` (`list of registers`)
+
+Specifies the list of vector registers allowed to be used for passing arguments and the order they are to be used in.
+
+## `VectorReturnValueRegisters` (`list of registers`)
+
+Specifies the list of vector registers allowed to be used for returning values and the order they are to be used in.
+
+## `CalleeSavedRegisters` (`list of registers`)
+
+Specifies the list of registers for which the ABI requires the callee to preserve the value, meaning that when the callee returns, the value of those registers must be the same as it was when the function was called.
+
+## `ReturnValueLocationRegister` (`register`)
+
+Specifies a register to be used for returning (or even passing, depending on ABI) the pointer to the memory used for returning copies of big aggregate objects.
+
+Must be set to `model::Register::Invalid` for ABIs that do not support returning values by 'pointer-to-copy' or use stack for it.
+
+## `ReturnValueLocationOnStack` (`bool`)
+
+Specifies whether stack is used to pass the return value location.
+
+This is only relevant if `ReturnValueLocationRegister` is set to `Invalid`.
+
+## `ReturnValueLocationIsReturned` (`bool`)
+
+Specifies whether functions whose return value is saved in a memory area pointed by an argument passed by the caller (SPTAR) should also return such pointer using the first return value register.
+
+This is only relevant if either `ReturnValueLocationRegister` or `ReturnValueLocationOnStack` is enabled.
+
+## `ScalarTypes` (`list of scalar types`)
```

Comment:
I think there's some notable example why we need this. A 32-bit integer with 64-bit alignment or viceversa, please report it.

Reply:
Nothing that drastic.

There are two factors that influenced this kind of list:

1. We wanted a very explicit list of supported scalars. So everything outside the list is out by definition.

2. The alignment is unpredictable, some ABI just say "everything is aligned at most at register size". Others are like "scalar size is the alignment no matter what", and all the steps in-between.

We could have probably got away with an enum / a couple of bools for the second point, but IMO, structuring it like this makes it a lot easier to read.